### ISSUE TITLE
Bug: play/pause toggle got out of sync with reality

### ISSUE THREAD

**Issue by mpogue2** [mpogue2/SquareDesk#290]
Regina reported that she had a case where the Play/Pause button was showing the PLAY icon, but the music was playing.  It should have been showing the PAUSE icon at that time.   This happened to be when her laptop had very low battery, and was feeling "slow" (indeed, Activity Monitor showed a lot of activity in terms of percent).  After plugging in power, we could not reproduce the problem.

I was never able to reproduce the problem on any laptops that I tried it on.  But, then again, all of my laptops were plugged in, and not under heavy CPU load.

That suggests a timing problem that only shows up when CPU load is high.

To help guard against this, we might take a belt and suspenders approach (if we can't quickly find the root cause).  For example, if cBass is playing, then infoBar could insist that the play/pause button is in the correct state.  It could also log the fact that this occurred (if we had a log facility)....

**Comment by mpogue2** [comment]
Regina saw this again (with me watching).  The laptop was again running on battery.  We saw TWO different symptoms:
1) When pressing period, the song position reset itself back to zero, after playing just for a few seconds.
2) Once, when pressing period (I think), the song bar kept going, but no music was coming out.

So, we've seen both ways:  stops playing when it should be playing, and playing (song bar moving) when it should not.

**Comment by mpogue2** [comment]
Failure was on 0.9.0 .  She has not seen any similar failures on 0.8.3 alpha7 .  Only seen on Regina's laptop, and only when it was on battery power (if that matters).

**Comment by mpogue2** [comment]
I reproduced this on a Release build (Top of Tree) on my machine at home.  Same symptom:  pressed "." and it would play for about 1 second and then stop.

Before this happened, I was quickly double clicking on a song, and then another song, back and forth.  It stopped in a state where I had just double clicked song B, but the title for song A was still showing.  Every time I clicked, it took a few seconds to update everything.

Once it got into this state, it couldn't get out by pressing the STOP button.  I had to quit and come back in.

So, I suspect that a) loading a song is too slow now (I don't know why), and b) if you quickly double click back and forth, it will get into an inconsistent state, and then "." won't work EVEN after waiting and double clicking on a song to load it properly.

This would be a race condition, I would guess.

**Comment by mpogue2** [comment]
Also note:  the songs I was clicking were singing calls, both loading lyrics.  This was also true of Regina's failure case -- we were practicing singing calls.

So, my guess is that the Lyrics editor adds load time, and double clicking from one song to another quickly is causing the inconsistent state.  This is probably much less likely to happen with patter, because patter just loads the default patter template, and they don't have anything in the dropdown menu on the Patter/Lyrics tab.

**Comment by danlyke** [comment]
Hmmm... I wonder if this is related to my calling laptop delay after
starting to play. I'll poke around what we do when we start playing and
see if there's anything I see as obviously extraneous.

**Comment by mpogue2** [comment]
I'm also thinking about looking at the ElapsedTimer again to see where the time is going, when Lyrics are loaded (unlike Patter, which seems to be a lot quicker).

**Comment by mpogue2** [comment]
Dan has a proposed fix for this, which involves changing some SQL stuff.  The commit is: bff84a4b0b71676d824d8e6fbbe96aeb413a4457 .  Looks like this does speed up the load song time, so even though we have not root-caused this problem, I'm hoping that this will reduce the incidence of the problem.

I was able to reproduce a case of this again, but only on the previous code, and only once, and only when I moved very quickly to double click, then period, then double click, etc.  I can't reproduce it right now, after this change, but that doesn't mean that the problem is gone.  It does still smell like a race condition in song loading.

**Comment by mpogue2** [comment]
OK I think I see how the failure can occur in theory.  The cBass thread might not have stopped yet, since stopping is likely asynchronous.  Then, since cBass.Play() was a toggle, it would have stopped playback, rather than starting it again.  I've separate the Play() into Play(), which is a true "PLAY NOW", and Pause(), which is a true "PAUSE NOW".  That should fix this problem.  Commit is 01d6c0c616768e31a00ac6c63b852afcab7c8137 .

It was a race condition, and now that I am shipping the RELEASE version, instead of the DEBUG version on Mac OS X, that might explain why this problem was seen NOW in 0.9.0, instead of in earlier releases.

At least, that's my story, and I'm sticking to it.  Let's reopen this bug, if the problem shows up again.

**Comment by mpogue2** [comment]
Reopening, because it's been spotted on two laptops (still not mine!).
Symptom on one laptop was play/pause was out of sync, then crash.

**Comment by mpogue2** [comment]
Regina saw this, but restarting the app made it go away.  It's hard to reproduce, but definitely still there.  She was also running very very low on battery when it happened, in case that might be a cause of the problem.  We suspected many weeks ago that battery was the problem then, but could not reproduce then either.  After putting her laptop on a charger, we were unable to reproduce in several hours of using SquareDesk.  I don't know whether battery is a red herring or not.

**Comment by mpogue2** [comment]
I've been able to reproduce this twice:
a) older laptop
b) battery allowed to drain to 12%
c) lots of start/stop using spacebar and period
d) about 1 failure out of every 50 tries
e) had touchpad active, with all 3 checkboxes in Trackpad checked (I don't know if this was required to create the problem).

When it failed, it was playing, but the icon on the play/pause button was PLAY (should have been PAUSE).  When I hit space bar, the icon changed to PAUSE, and it kept playing.

Double-clicking the infoBar had no effect when it was in this mode.  Just kept playing.
The debug.log file does NOT show that the belt and suspenders code was triggered.
Song was Amie + BGVs.

**Comment by mpogue2** [comment]
Before I finished last night, I was able to get one failure at 62% battery.  So, I think the problem is maybe always was there, and the battery idea was coincidence.  Still maybe 1 failure out of 50 tries, and only on one of my 2 laptops.

I'm going to try a different belt/suspenders approach to see if it helps.

**Comment by mpogue2** [comment]
OK, I think this one is fixed now.  I got rid of ALL cached versions of the libbass engine state (Stream_State and currentState), and if it's needed, we always ask the engine on-the-fly now.  Previous failures were 1/50 tries on old laptop.  Now, I can't get it to fail out of 500 tries on same laptop.  Hopefully gone for good.

The underlying problem was that I was keeping cached libbass engine state in a handy-to-use and cheap-to-read local variable.  But, there were tiny windows where the cached value didn't exactly match the state of the engine, since the engine is asynchronous to my poking at it.  Fix is to always ask the engine what state it's in.  It turns out that it's cheap enough to ask every time.

Fixed by 41019c564fc0bf2745338a68c3162e0888ff67e9 .
